modules/cudastereo/src/cuda/stereosgm_utility.hpp

```diff
@@ -0,0 +1,227 @@
+/*
+Copyright 2016 Fixstars Corporation
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http ://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#ifndef SGM_UTILITY_HPP
+#define SGM_UTILITY_HPP
+
+#include <cuda.h>
+
+namespace cv { namespace cuda { namespace device
+{
+    namespace stereosgm
+    {
+        static constexpr unsigned int WARP_SIZE = 32u;
+
+        namespace detail {
+            template <typename T, unsigned int GROUP_SIZE, unsigned int STEP>
+            struct subgroup_min_impl {
```

Comment:
I think that we have to use [cudev::warpReduce](https://github.com/opencv/opencv_contrib/blob/c8ed08d0a245656c75617dfc93a9f5b6b84aaf8f/modules/cudev/include/opencv2/cudev/warp/reduce.hpp#L62).

Reply:
cudev::warpReduce calls __shfl_down internally and not supports mask.
It made codes which broadcast reduced value necessary.

Additionally, `core` module define similar utility functions,
so we need not to depend `cudev` module.

Reply:
> Additionally, core module define similar utility functions,

Sry, cuda utility in `core` module is deprecated.

https://github.com/opencv/opencv/blob/1f517b8a0293f2574c3dce739e5b273164244cac/modules/core/include/opencv2/core/cuda/warp_shuffle.hpp#L47
> `@deprecated Use @ref cudev instead.`